cvat/apps/engine/models.py

```diff
@@ -181,6 +183,242 @@ def choices(cls):
     def __str__(self):
         return self.value
 
+
+T = TypeVar("T")
+
+
+def _parse_both_before_accessing(fn):
+    @wraps(fn)
+    def wrapper(self: 'LazyList', other: Any) -> 'LazyList':
+        self._parse_up_to(-1)
+        if not isinstance(other, list):
+            # explicitly calling list.__add__ with
+            # np.ndarray raises TypeError instead of it returning NotImplemented
+            # this prevents python from executing np.ndarray.__radd__
+            return NotImplemented
+        if isinstance(other, LazyList):
+            other._parse_up_to(-1)
+
+        return fn(self, other)
+
+    return wrapper
+
+
+def _parse_before_accessing(fn: Callable[..., Any]) -> Callable[..., Any]:
+    """Wrapper for original list methods. Forces LazyList to parse itself before accessing them."""
+    @wraps(fn)
+    def wrapper(self: 'LazyList', *args, **kwargs) -> 'LazyList':
+        self._parse_up_to(-1)
+
+        return fn(self, *args, **kwargs)
+
+    return wrapper
+
+
+class LazyList(list[T]):
+    """
+    Evaluates elements from the string representation as needed.
+    Lazy evaluation is supported for __getitem__ and __iter__ methods.
+    Using any other method will result in parsing the whole string.
+    Once instance of LazyList is fully parsed (either by accessing list methods
+    or by iterating over all elements), it will behave just as a regular python list.
+    """
+    __slots__ = ("string", "_separator", "_converter", "_probable_length", "parsed")
+    string: str
+    _separator: str
+    _converter: Callable[[str], T]
+    _probable_length: int | None
+
+    def __init__(self, string: str = "", separator: str = ",", converter: Callable[[str], T] = lambda s: s) -> None:
+        super().__init__()
+        self.string = string
+        self._separator = separator
+        self._converter = converter
+        self._probable_length = None
+        self.parsed = False
```

Comment:
I think, you can reduce the amount of code here by using `attrs.define` or `dataclass`.

Reply:
I went with `dataclass` initially, however I think it's best to keep these attributes underscored, yet keep `__init__` arguments public (i.e. `LazyList('1,2,3', converter=int)` instead of `LazyList('1,2,3', _converter=int)`. 

AFAIK, there's no simple way to allow that with `dataclass` or `attr.define`.

Reply:
For attrs, it's explained [here](https://www.attrs.org/en/stable/examples.html#basics) ("For private attributes, ..."). Basically, it's the default behavior for underscored fields.